### ISSUE TITLE
Better STIX Parsing

### ISSUE THREAD

**Issue by alatif113** [MISP/MISP-STIX-Converter#23]
Currently every observable is parsed out of a given STIX input and added as an attribute to a MISP event. In many cases this can give undesired results. A better solution would be as follows:

1.  Create an observable map (python dictionary) with the observable id as a key and the observable object as the value, for every observable within the STIX package. 
2. Create a TTP map (python dictionary) with the TTP id as a key and the TTP object as the value, for every TTP within the STIX package. 
3. If there are _indicators_ within the STIX package, parse the indicator. Use the observable map to retrieve observables part of that indicator (typically they'll be represented only by the ID within the indicator). Use the TTP map to retrieve TTPs part of that indicator (typically they'll be represented only by the ID within the indicator). 

- Parsing indicators, rather than straight observables will give you the added value of confidence levels, TTPs, campaigns, actors, etc that can possibly be added as tags to the MISP attributes. 
- When parsing indicators, you can ignore observables that are part of an "AND" composition to prevent unwanted results such as a ".exe" file extension attribute.  These observables are part of a group and do not make sense on their own. It's best to ignore them until there's functionality within MISP for it. 

4. If no indicators exists (some simple feeds will not use them), you can fall back to parsing straight indicators. Just loop through the observable map you created before.

**Comment by alatif113** [comment]
@FloatingGhost I added sample code but the formatting is not cooperating. Can you help me fix it please?

**Comment by iglocska** [comment]
Sounds like some very sane ideas!

**Comment by FloatingGhost** [comment]
o christ on a quadbike that's a lot of edited code

I might take a look when I'm feeling more masochistic than usual

**Comment by alatif113** [comment]
@FloatingGhost lol most of it is just the parsing of the different observable types taken directly from the existing code.

**Comment by FloatingGhost** [comment]
>just

```
JUST
```

Nothing with STIX is ever "JUST" 

It's always rooted to the hellish floor of the so-called standard and is never as simple as it seems. Inevitably there'll be a tendril of this eldritch horror that extends beyond the realm of human comprehension and into STIX world, in which it terminates in the STIXPackage of unimaginable terror

**Comment by iglocska** [comment]
@alatif113, could you do a pull request for the changes? It does sound like a clean approach indeed.

**Comment by alatif113** [comment]
@iglocska I'm not very git savvy. Don't really know how to do that.

**Comment by iglocska** [comment]
Should be simple enough, simply fork the project by clicking the "Fork" button on top, this will create a copy of the repository under your user name on github, so https://github.com/alatif113/MISP-STIX-Converter

You already have the code-base sitting locally on your machine, which is what you have modified. Create a commit with all of your changes by simply doing the following:

```
git add /path/to/your/changed/file
```
Repeat it for all of the files that you have modified (to see a list of all files that you've changed just type ```git status``` from within the MISP-STIX-Converter directory)

Once you are done it's time to commit the changes:

```
git commit -m "My STIX parsing improvements"
```

Once done, add your own github repository as a remote 

```
git remote add myfork https://github.com/alatif113/MISP-STIX-Converter.git
```

Then push your committed changes to your fork: 

```
git push myfork master
```

Once this is done, just go to 

```
https://github.com/alatif113/MISP-STIX-Converter
```
and open up a pull request by clicking the "New pull request" button (upper left side, next to the branch name)

This should be it!

**Comment by alatif113** [comment]
@iglocska Ahh I see what you mean. I didn't edit any files directly (I didnt need the MISP to STIX part), but rather created my own 2 files (in the code above) solely for parsing and importing STIX to MISP, heavily using code that already existed within the project to parse the actual observable types.

**Comment by iglocska** [comment]
Ah ok, I see. Any chance you could move that to MISP-STIX-Converter and integrate it directly? Or is it too different from how the converter works?

**Comment by alatif113** [comment]
@iglocska I think replacing the `buildEvent` function within the `buildMISPAttribute` file with the build_event function within the `stix_to_misp.py` file above should do it, barring the fact there would now be a lot of unused old functions.

There would also need to be some agreement on the taxonomy for tags. I use `Confidence:<value>` and `TTP:<value>` for mine, but I don't know if there is already some standard that exists out there.  

Unfortunately I don't have the time to actually go through and do that. Just wanted to bring something to the dev's attention with sample code I utilized for my use case.

**Comment by iglocska** [comment]
Thanks a lot for the input, we'll keep this issue open until we can get around to implementing it. It indeed looks very promising!

**Comment by alatif113** [comment]
@iglocska No problem! Got the idea from how many of the SIEMs and commercial threat platforms parse STIX files.

**Comment by FloatingGhost** [comment]
Well if they parsed MISP everyone's life would be a lot easier :<

STIX needs to die.

**Comment by iglocska** [comment]
It looks like it's here to stay though, so we should make sure that the parser makes as much sense as possible - we'll definitely take a look at this too at some point, @FloatingGhost, to preserve some of your sanity ;))

**Comment by safianouyacouba** [comment]
> **Slu

**Comment by ag-michael** [comment]
Any updates on this? @iglocska , it seems you approve of the general idea and there is demand for this feature. Any chance of accepting PR's related to this soon?

**Comment by ag-michael** [comment]
@alatif113 I have a Pending PR https://github.com/MISP/MISP-STIX-Converter/pull/40 that addresses some of what you're wanting, care to take a look and comment? This is an important subject for me as well.

**Comment by adulau** [comment]
So we don't really maintain this as there is a full-blown STIX 1.x and 2.x import/export in MISP. 

As the original maintainer is not taking part of this external package, I can merge those. Just let me know if it works for you and I'll merge it.

**Comment by ag-michael** [comment]
@adulau it would be great if you can merge it. But I didn't know about the MISP feature for STIX import, do you know why MISP-Taxii-Server isn't using that? That's the only way to feed MISP with TAXII that I've found, can you point me in the direction of the docs that show how to import STIX directly into MISP
